### PR TITLE
Upgrade ESLint and add more rules

### PR DESCRIPTION
**Pull request by ekowidianto**
### Changes:
- Upgraded eslint to v8.

- Added various new plugins as follows:
   - eslint-plugin-sonarjs - for bugs and suspicious patterns detection.
   - eslint-plugin-eslint-comments - for directive comments.
   - eslint-plugin-simple-import-sort - for lib import sorting.

### Note:
If you get the following error:
```
Parsing error: No Babel config file detected for `filename`. Either disable config file checking with requireConfigFile: false, or configure Babel so that it can find the config files.eslint
```
Add the following to the vscode workspace setting config
```    
"eslint.workingDirectories": [
        {"mode": "auto"}
    ],
```

Source: `https://github.com/microsoft/vscode-eslint/issues/1397`

### CODE REVIEW

**Comment by ekowidianto** [comment]
@purfectliterature Any other inputs for the linter are welcomed! Could you also rebase all your future PRs off this branch? Thanks

File: client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx [hunk 1]
```diff
@@ -25,19 +26,19 @@ const ZoomActionElements = (props) => {
   return (
     <div style={styles.wrapper}>
       <Button
-        label="Zoom In"
-        color="primary"
         className="btn-submit"
+        color="primary"
+        label="Zoom In"
```
Comment (purfectliterature) [inline client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx:31]: Extract to `translations.intl.js`?

File: client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx [hunk 2]
```diff
@@ -25,19 +26,19 @@ const ZoomActionElements = (props) => {
...
         variant="contained"
       >
         <FormattedMessage {...translations.zoomIn} />
       </Button>
       <Button
-        label="Zoom Out"
-        color="primary"
         className="btn-submit"
+        color="primary"
+        label="Zoom Out"
```
Comment (purfectliterature) [inline client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx:41]: Extract to `translations.intl.js`?

File: client/app/bundles/course/video/pages/VideoShow/index.tsx [hunk 3]
```diff
@@ -79,24 +82,22 @@ const VideoShow: FC<Props> = (props) => {
...
+  ) : null;
   const returnLink = video?.tabId
     ? `${getVideosURL(getCourseId())}?tab=${video.tabId}`
     : getVideosURL(getCourseId());
   return (
     <>
       <PageHeader
-        title={`Video - ${video?.title}`}
         returnLink={returnLink}
+        title={`Video - ${video?.title}`}
```
Comment (purfectliterature) [inline client/app/bundles/course/video/pages/VideoShow/index.tsx:100]: Extract to translations file?

File: client/app/bundles/course/video/submission/pages/VideoSubmissionShow/index.tsx [hunk 4]
```diff
@@ -88,8 +91,8 @@ const VideoSubmissionShow: FC<Props> = (props) => {
     return (
       <>
         <PageHeader
-          title="Video"
           returnLink={getVideoSubmissionsURL(getCourseId(), getVideoId())}
+          title="Video"
```
Comment (purfectliterature) [inline client/app/bundles/course/video/submission/pages/VideoSubmissionShow/index.tsx:95]: Extract to translations?

File: client/app/bundles/course/video/submission/pages/VideoSubmissionShow/index.tsx [hunk 5]
```diff
@@ -111,23 +114,21 @@ const VideoSubmissionShow: FC<Props> = (props) => {
...
-  ) : (
-    <></>
-  );
+  ) : null;
   return (
     <>
       <PageHeader
-        title={`Video - ${videoSubmission?.videoTitle}`}
         returnLink={getVideoSubmissionsURL(getCourseId(), getVideoId())}
+        title={`Video - ${videoSubmission?.videoTitle}`}
```
Comment (purfectliterature) [inline client/app/bundles/course/video/submission/pages/VideoSubmissionShow/index.tsx:131]: Extract to translations?

File: client/app/bundles/system/admin/admin/components/buttons/InstancesButtons.tsx [hunk 6]
```diff
@@ -55,24 +57,22 @@ const InstancesButtons: FC<Props> = (props) => {
...
-          disabled={isDeleting}
-          loading={isDeleting}
-          onClick={onDelete}
           confirmMessage={intl.formatMessage(translations.deletionConfirm, {
             name: instance.name,
           })}
+          disabled={isDeleting}
+          loading={isDeleting}
+          onClick={onDelete}
+          tooltip="Delete Instance"
```
Comment (purfectliterature) [inline client/app/bundles/system/admin/admin/components/buttons/InstancesButtons.tsx:71]: Extract to translations?

**Review by purfectliterature — CHANGES_REQUESTED** [review]
Thanks for adding these useful rules! We might still be miles away from a clean, neat codebase, but these definitely are steps in the right direction.

File: client/app/bundles/course/assessment/question/programming/actions/programmingQuestionActionCreators.jsx [hunk 7]
```diff
@@ -1,6 +1,9 @@
 import axios from 'axios';
+
 import actionTypes from '../constants/programmingQuestionConstants';
 
+const headerType = 'application/json';
```
[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/bundles/course/assessment/question/programming/actions/programmingQuestionActionCreators.jsx:5]: Make sense! It's a js file so I wont include `const`

File: client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx (hunk 1 above)
Comment (ekowidianto) [reply client/app/bundles/course/learning-map/components/ZoomActionElements/index.jsx:31]: `yarn lint-fix` rearranges all the props in this PR. The initial purpose of this PR is to fix the linting, and not extract the intl. I suspect there's other labels that are not intl-ed yet. I will fix the ones highlighted in the comment but maybe another new issue should be opened to properly go through all the code and extract the labels.

File: client/.eslintrc.js [hunk 8]
```diff
@@ -34,12 +53,46 @@ module.exports = {
...
     'react/destructuring-assignment': 'off',
     'react/forbid-prop-types': ['error', { forbid: ['any', 'array'] }],
+    'react/function-component-definition': [
+      'error',
+      {
+        namedComponents: 'arrow-function',
+        unnamedComponents: 'arrow-function',
+      },
+    ],
+    'react/jsx-boolean-value': ['error', 'always'],
```
[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/.eslintrc.js:63]: Good points! Thanks for highlighting this

**Review by purfectliterature — APPROVED** [review]
Let's go!

**Pull request merged**